core/common/src/main/java/alluxio/util/ConfigurationUtils.java

```diff
@@ -413,7 +413,12 @@ public static void reloadProperties() {
       // property file.
       AlluxioProperties properties = new AlluxioProperties();
       InstancedConfiguration conf = new InstancedConfiguration(properties);
-      properties.merge(System.getProperties(), Source.SYSTEM_PROPERTY);
+      // Can't directly pass System.getProperties() because it is not thread-safe
+      // This can cause a ConcurrentModificationException when merging.
+      Properties sysProps = new Properties();
+      System.getProperties().stringPropertyNames()
+          .forEach(key -> sysProps.setProperty(key, System.getProperty(key)));
```

Comment:
Can this also throw concurrent modification exception?

Reply:
No, The set returned from `stringPropertyNames` is a copy of the keys at the time of calling:

From: https://docs.oracle.com/javase/8/docs/api/java/util/Properties.html#stringPropertyNames--

>  The returned set is not backed by the Properties object. Changes to this Properties are not reflected in the set, or vice versa.

https://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/f54e9b7c1036/src/share/classes/java/util/Properties.java#l1030

Reply:
The implementation of `stringPropertyNames` uses an enumerator on the keys?

Reply:
The enumerator is thread safe. Calling `stringPropertyNames` loops over [the `keys()` method of the Hashtable](https://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/f54e9b7c1036/src/share/classes/java/util/Properties.java#l1102)

- [`keys()`](https://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/f54e9b7c1036/src/share/classes/java/util/Hashtable.java#l248) is implemented via an internal [`Enumeration` object](https://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/f54e9b7c1036/src/share/classes/java/util/Hashtable.java#l1316)

Iterating over all of the keys with the `Enumerator` is safe against `ConcurrentModificationException` so long as `remove()` and `next()` isn't called. Neither are called because the Properties implementation uses [`nextElement()` rather than `next()`](https://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/f54e9b7c1036/src/share/classes/java/util/Properties.java#l1097)